binding.c

```diff
@@ -287,38 +287,71 @@ bare_tcp__on_close(uv_handle_t *handle) {
   err = js_get_reference_value(env, tcp->ctx, &ctx);
   assert(err == 0);
 
-  js_value_t *on_close;
-  err = js_get_reference_value(env, tcp->on_close, &on_close);
-  assert(err == 0);
+  if (tcp->resetting) {
+    uv_loop_t *loop;
+    err = js_get_env_loop(env, &loop);
+    assert(err == 0);
 
-  err = js_delete_reference(env, tcp->on_connection);
-  assert(err == 0);
+    uv_tcp_t new_handle;
+    tcp->handle = new_handle;
 
-  err = js_delete_reference(env, tcp->on_connect);
-  assert(err == 0);
+    err = uv_tcp_init(loop, &tcp->handle);
 
-  err = js_delete_reference(env, tcp->on_read);
-  assert(err == 0);
+    if (err < 0) {
+      js_throw_error(env, uv_err_name(err), uv_strerror(err));
+      return;
+    }
 
-  err = js_delete_reference(env, tcp->on_write);
-  assert(err == 0);
+    js_value_t *on_reset;
+    err = js_get_reference_value(env, tcp->on_reset, &on_reset);
+    assert(err == 0);
 
-  err = js_delete_reference(env, tcp->on_end);
-  assert(err == 0);
+    js_call_function(env, ctx, on_reset, 0, NULL, NULL);
 
-  err = js_delete_reference(env, tcp->on_close);
-  assert(err == 0);
+    err = js_close_handle_scope(env, scope);
+    assert(err == 0);
 
-  err = js_delete_reference(env, tcp->ctx);
-  assert(err == 0);
+    tcp->resetting = false;
```

Comment:
This must happen before `js_call_function()` as the we no longer own the `tcp` handle after passing control back to JavaScript.